src/condition/ssrf_detector.cpp

```diff
@@ -0,0 +1,279 @@
+    "10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16", "100.64.0.0/10", "::1/128", "fc00::/7",
+    "fe80::/10", "2001:db8:1234:1a00::/56"};
+
+constexpr const std::array<std::string_view, 10> dangerous_domains{
+    "metadata.google", "burpcollaborator.net", ".local", ".internal", "ram.aliyuncs.com",
+    "ifconfig.pro", "localhost", "fuf.me", "localtest.me", "ulh.us" // Legacy domains
+};
+
+constexpr const std::array<std::string_view, 4> authorised_schemes{"https", "http", "ftps", "ftp"};
+
+constexpr const auto &npos = std::string_view::npos;
+
+using ssrf_result = std::optional<std::pair<std::string, std::vector<std::string>>>;
+
+bool detect_parameter_injection(
+    const uri_decomposed &uri, std::string_view param, std::size_t param_index)
+{
+    const auto path_end = uri.path_index + uri.path.size();
+
+    // Check if the application is giving full control of the path to the user
+    // either fully or after a forward-slash.
+    if (!uri.path.empty()) {
+        //  scheme://userinfo@host:port/path?query#fragment
+        //                             ─────
+        if (uri.path.size() == param.size()) {
+            return false;
+        }
+        //  scheme://userinfo@host:port/path?query#fragment
+        //                              ────
+        if (uri.path[0] == '/' && param_index == uri.path_index + 1 &&
+            param.size() == uri.path.size() - 1) {
+            return false;
+        }
+    }
+
+    // REST parameter injection will involve introducing a / to the URL, however, such
+    // slashes are allowed after the ? or the #. Check if the slash is within the path
+    //
+    //  scheme://userinfo@host:port/path?query#fragment
+    //                             ─────
+    if (auto slash_index = param.find('/'); slash_index != npos) {
+        slash_index += param_index;
+        if (slash_index < path_end) {
+            // The path is partially under control of the user, this might be intentional
+            // so lets check for a possible LFI
+            auto relative_dir_index = param.find("..");
+            while (relative_dir_index != npos) {
+
+                // We found '..', check if it's enclosed by '/'
+                auto dir_index = relative_dir_index + param_index;
+                if (dir_index < path_end && (dir_index + 2) < uri.raw.size() &&
+                    uri.raw[dir_index - 1] == '/' && uri.raw[dir_index + 2] == '/') {
+                    return true;
+                }
+                relative_dir_index = param.find("..", relative_dir_index + 2);
+            }
+        }
+    }
+
+    // The application giving total control to the user after a given point.
+    // If it starts in the path, assume it's intentional
+    //
+    //  scheme://userinfo@host:port/path?query#fragment
+    //                             ────────────────────
+    if ((uri.query.empty() && uri.fragment.empty()) ||
+        (param_index < path_end && (param_index + param.size()) == uri.raw.size())) {
```

Comment:
how does this match the comment? FWICT, the check after the disjunction is if the param starts before path_end (and then the param finishes at the end of the raw url), not if it starts in the same place the path starts in the uri

Reply:
It just refers to the fact that everything after a certain point has been injected, which might actually be the intention of the developer. I can clarify it.